core/cppgc.rs

```diff
@@ -18,6 +18,80 @@ struct CppGcObject<T: GarbageCollected> {
   member: T,
 }
 
+/// PrototypeChain defines the position of the object in the prototype chain.
+///
+///   A    (index 0)
+///  / \
+/// B   C  (index 1)
+/// |
+/// D      (index 2)
+///
+/// where each B, C and D hold their own prototype chain.
+pub trait PrototypeChain {
+  fn prototype_index() -> Option<usize> {
+    None
+  }
+}
+
+/// A good enough number for the maximum prototype chain depth based on the
+/// usage.
+const MAX_PROTO_CHAIN: usize = 3;
+
+struct DummyT;
+impl GarbageCollected for DummyT {}
+
+/// ErasedPtr is a wrapper around a `v8::cppgc::Ptr` that allows downcasting
+/// to a specific type. Safety is guaranteed by the `tag` field in the
+/// `CppGcObject` struct.
+struct ErasedPtr {
+  ptr: v8::cppgc::Ptr<CppGcObject<DummyT>>,
+}
+
+impl ErasedPtr {
+  fn downcast<T: GarbageCollected + 'static>(
+    &self,
+  ) -> Option<v8::cppgc::Ptr<CppGcObject<T>>> {
+    if self.ptr.tag == TypeId::of::<T>() {
+      // Safety: The tag is always guaranteed by `wrap_object` to be the
+      // correct type.
+      Some(unsafe { std::mem::transmute_copy(&self.ptr) })
+    } else {
+      None
+    }
+  }
+}
+
+impl<T: GarbageCollected> From<v8::cppgc::Ptr<CppGcObject<T>>> for ErasedPtr {
+  fn from(ptr: v8::cppgc::Ptr<CppGcObject<T>>) -> Self {
+    debug_assert!(
+      std::mem::size_of::<v8::cppgc::Ptr<CppGcObject<T>>>()
+        == std::mem::size_of::<ErasedPtr>()
+    );
+
+    Self {
+      // Safety: Both have the same size, representation and a tag guarantees safe
+      // downcasting.
+      ptr: unsafe {
+        std::mem::transmute::<
+          v8::cppgc::Ptr<CppGcObject<T>>,
+          v8::cppgc::Ptr<CppGcObject<DummyT>>,
+        >(ptr)
+      },
+    }
+  }
+}
+
+struct PrototypeChainStore([Option<ErasedPtr>; MAX_PROTO_CHAIN]);
+
+impl v8::cppgc::GarbageCollected for PrototypeChainStore {
+  fn trace(&self, visitor: &v8::cppgc::Visitor) {
+    // Trace all the objects top-down the prototype chain.
+    for ptr in self.0.iter().flatten() {
+      ptr.ptr.trace(visitor);
```

Comment:
Does this dispatch correctly? We don't want to call `EreasedPtr`'s T here - I assume this all works out, but can you add some comments?